feat: Rework sslyze parser for version 5

kibernautas opened the pull request:
The sslyze parser changed schema when it updated to v5, so I have rewritten the parser partially in order to parse the JSON file. 

- [x] rework parser for v5
- [x] create new unit test cases for v5
- [x] create fallback scan option for v3

damiencarol commented:
@kibernautas you will need to update unit tests also.

valentijnscholten commented:
also we may want to make it backwards compatible. i.e. support the older format as well.

StefanFl commented:
Hi @kibernautas, are you still working on it? It would be great to have the changes in the next release.

kibernautas commented:
Hi @StefanFl I was a bit burdened with work so I will continue this next week if that's okay, I assume that from the tips that you guys provided I need to update the json files for sslyze for the unit tests and implement another scan option for the older version of sslyze? If I am missing something please do correct me.

StefanFl commented:
Great news 😃 Have a look at https://github.com/DefectDojo/django-DefectDojo/blob/master/dojo/tools/gitleaks/parser.py, this is an example of a parser which supports 2 versions of JSON files.

kibernautas commented:
Welp, I added an if to check for the JSON patterns and from there on out it will decide on what to parse. For the test cases, I have kept the old JSON files and added new ones, so it should pass with both of them.

mtesauro commented:
If you don't mind, could you do a PR of the new file format into this Sample Files repo aka https://github.com/DefectDojo/sample-scan-files/tree/master/sslyze  

It looks like the ones we have are 3 years old.

Note: Only submit sanitized sample files where there's no internal/private information included. Most people scan something that is public or open source.

Thanks!

kibernautas commented:
I have added the files, replaced the domain names with example.com 
https://github.com/DefectDojo/sample-scan-files/pull/79

mtesauro commented:
> I have added the files, replaced the domain names with example.com [DefectDojo/sample-scan-files#79](https://github.com/DefectDojo/sample-scan-files/pull/79)

Already merged.  Thanks again.

kibernautas commented:
Happy to help

StefanFl reviewed (changes requested):
Two things before I start reading the code:

- I don't know what you have changed in `dojo/fixtures/defect_dojo_sample_data.json` because GitHub can't render the changes. But this file shouldn't be affected anyway by this PR. Please revert your changes.
- The JSON scan files for the unit tests in `unittests\scans\sslyze` are missing, that's why the unit tests are not working.

kibernautas commented:
> Two things before I start reading the code:
> 
> * I don't know what you have changed in `dojo/fixtures/defect_dojo_sample_data.json` because GitHub can't render the changes. But this file shouldn't be affected anyway by this PR. Please revert your changes.
> * The JSON scan files for the unit tests in `unittests\scans\sslyze` are missing, that's why the unit tests are not working.

In the fixtures I changed the SSlyze names but I have reverted it, and added the were missing, hope it goes okay now.

damiencarol commented:
@kibernautas please skip the fixture file, it's not needed for this PR. Also changing the name of a parser is a big task, to stay aligned with other parser you will need to create a DB migration if you change the id a a parser.
Last point, please add the new version reports for new unit tests.

damiencarol commented on dojo/tools/sslyze/parser.py:
```diff
@@ -6,14 +6,14 @@ class SslyzeParser(object):
 
     def get_scan_types(self):
-        return ["SSLyze 3 Scan (JSON)", "Sslyze Scan"]
+        return ["Sslyze Scan", "SSLYze Scan(JSON)"]
```
As you are changing the scan_type value, we will need a DB migration.

StefanFl replied:
I am afraid to say I caused the problem initially, when I set the scan type to "SSLyze 3 Scan (JSON)". Now we have SSLyze5. I think we have to change the scan type. Tomorrow I will do a test, what's the best way to do it.

StefanFl replied:
Yes, we will need a DB migration if we change the scan type, otherwise reimports are not possible. While testing the reimport I saw a problem with the deduplication settings I want to investigate. 

@kibernautas Do you mind if I do these 2 issues in your branch?

kibernautas replied:
Not at all be my guest :D

damiencarol commented on unittests/tools/test_sslyze_parser.py:
```diff
@@ -6,40 +6,73 @@
+        for finding in findings:
+            for endpoint in finding.unsaved_endpoints:
+                endpoint.clean()
+        self.assertEqual(2, len(findings))
```
Please add more checks on the findings, like tests on the endpoints etc... you can take a look at other parsers (like Burp one) to have an example

kibernautas replied:
Will do that in the end of the week if that is okay.

damiencarol reviewed (changes requested):
Please revert changes in dojo/fixtures/defect_dojo_sample_data.json

StefanFl commented:
The rename of the parser is finished with a db migration and upgrade notes. Furthermore the deduplication settings have been fixed and I reverted the formatting change of the fixture.

damiencarol commented:
@kibernautas  @StefanFl  this is not ready. many changes should not be in the PR (like the modifications on the Docker script entry point).
Also why changing some test files of unit tests?
The goal of some fixes was to make the parser more robust and don't crash when some data are not there.
So it's better to add new reports where the data are not there instead of modifying old tests.

StefanFl commented:
Yes, it’s not finished yet. I just wanted to push what I have done. The change of the entry point was not invented, I will reverse that.

I didn’t change any previously existing files. But the new test scans where not suitable. The file that is supposed to have no vulnerabilities had 4 vulnerabilities. These files need to be changed to achieve the desired results.




Stefan Fleckenstein
Bereichsleiter Cybersecurity


Mobil +49 151 544 22 035
***@***.***
Theresienhöhe 13, 80339 München

MaibornWolff GmbH, Theresienhöhe 13, 80339 München
www.maibornwolff.de, Telefon +49 89 544 253 000
USt-ID DE 129 299 525, Amtsgericht München HRB 98058
Geschäftsführer: Volker Maiborn, Holger Wolff,
Alexander Hofmann, Florian Theimer, Marcus Adlwart

_______________________________________________

On 27 Jan 2022, at 19:28, Damien Carol ***@***.***> wrote:

﻿

@kibernautas<https://smex-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fkibernautas&umid=17ebcad0-d8f2-437f-9a7e-1edea0d3d9cc&auth=ba40f14edb9fac94ddb78b906f7486a30b35689c-2d96a801b79a9207ae9ecda7778d3d907bdedaf8> @StefanFl<https://smex-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fStefanFl&umid=17ebcad0-d8f2-437f-9a7e-1edea0d3d9cc&auth=ba40f14edb9fac94ddb78b906f7486a30b35689c-3ceea32b154f371da1690847818538f2605c10d3> this is not ready. many changes should not be in the PR (like the modifications on the Docker script entry point).
Also why changing some test files of unit tests?
The goal of some fixes was to make the parser more robust and don't crash when some data are not there.
So it's better to add new reports where the data are not there instead of modifying old tests.

—
Reply to this email directly, view it on GitHub<https://smex-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fDefectDojo%2fdjango%2dDefectDojo%2fpull%2f5689%23issuecomment%2d1023566079&umid=17ebcad0-d8f2-437f-9a7e-1edea0d3d9cc&auth=ba40f14edb9fac94ddb78b906f7486a30b35689c-c57d4cf54593c1a98f31bf3ca33d23bf3fdb102f>, or unsubscribe<https://smex-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fnotifications%2funsubscribe%2dauth%2fAAUS2BXW7F5BXSZF2PIMMPLUYGMD3ANCNFSM5LKDQ57Q&umid=17ebcad0-d8f2-437f-9a7e-1edea0d3d9cc&auth=ba40f14edb9fac94ddb78b906f7486a30b35689c-b7c822233448a9391eecef529c3f3b8430d01d54>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

damiencarol commented:
if it's a wrong name, it's ok, just rename the file. but please don't modify by hand the reports, it just mean that the unit tests should be modified with the code.
Also if we need an empty report most of the time, it's easy to produce a new one.

StefanFl commented:
Quite a lot of work to setup several sites with different TLS configurations to generate the reports. Won’t have a chance to do that but would like to support the new SSLyze version next week.




Stefan Fleckenstein
Bereichsleiter Cybersecurity


Mobil +49 151 544 22 035
***@***.***
Theresienhöhe 13, 80339 München

MaibornWolff GmbH, Theresienhöhe 13, 80339 München
www.maibornwolff.de, Telefon +49 89 544 253 000
USt-ID DE 129 299 525, Amtsgericht München HRB 98058
Geschäftsführer: Volker Maiborn, Holger Wolff,
Alexander Hofmann, Florian Theimer, Marcus Adlwart

_______________________________________________

On 27 Jan 2022, at 20:40, Damien Carol ***@***.***> wrote:

﻿

if it's a wrong name, it's ok, just rename the file. but please don't modify by hand the reports, it just mean that the unit tests should be modified with the code.
Also if we need an empty report most of the time, it's easy to produce a new one.

—
Reply to this email directly, view it on GitHub<https://smex-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fDefectDojo%2fdjango%2dDefectDojo%2fpull%2f5689%23issuecomment%2d1023620288&umid=44dd18e7-5fc6-4631-8727-e437cfc1ac2b&auth=ba40f14edb9fac94ddb78b906f7486a30b35689c-521e8fcec4a16046aa2ef68cd4fb13f8d3f3cee1>, or unsubscribe<https://smex-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fnotifications%2funsubscribe%2dauth%2fAAUS2BQCSEMEIMGLU437CUTUYGUS7ANCNFSM5LKDQ57Q&umid=44dd18e7-5fc6-4631-8727-e437cfc1ac2b&auth=ba40f14edb9fac94ddb78b906f7486a30b35689c-7c34b2a53865468940e074d1f515f0b902427ae1>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
